sw/device/sca/aes_serial/aes_serial.c

```diff
@@ -0,0 +1,334 @@
+ */
+static void print_hex_buffer(const char *data, size_t data_len) {
+  static const char b2a_hex_values[16] = "0123456789ABCDEF";
+  for (int i = 0; i < data_len; ++i) {
+    uart_send_char(b2a_hex_values[data[i] >> 4]);
+    uart_send_char(b2a_hex_values[data[i] & 0xF]);
+  }
+}
+
+/**
+ * Send simple serial command response via UART.
+ *
+ * @param cmd_tag  Simple serial command tag.
+ * @param data     Response data. Converted to hex format by this function.
+ * @param data_len data length.
+ */
+static void print_cmd_response(const char cmd_tag, const char *data,
+                               size_t data_len) {
+  uart_send_char(cmd_tag);
+  print_hex_buffer(data, data_len);
+  uart_send_char('\n');
+}
+
+/**
+ * Return number of AES blocks from `plain_text_len`.
+ */
+static size_t get_num_blocks(size_t plain_text_len) {
+  size_t num_blocks = plain_text_len / kAesBlockSize;
+  if ((plain_text_len - (num_blocks * kAesBlockSize)) > 0) {
+    ++num_blocks;
+  }
+  return num_blocks;
+}
+
+/**
+ * Simple serial set AES key command.
+ *
+ * @param aes_cfg AES configuration object.
+ * @param key     AES key.
+ * @param key_len AES key length.
+ *
+ * @return kSimpleSerialOk on success, kSimpleSerialError otherwise.
+ */
+static simple_serial_result_t simple_serial_set_key(const aes_cfg_t *aes_cfg,
+                                                    const char *key_share0,
+                                                    size_t key_len) {
+  // The implementation does not use key shares to simplify AES attacks.
+  static const uint8_t kKeyShare1[32] = {0};
+
+  if (key_len != kAesKeySize) {
+    return kSimpleSerialError;
+  }
+
+  aes_clear();
+  while (!aes_idle()) {
+    ;
+  }
+  aes_init(*aes_cfg);
+
+  aes_key_put(key_share0, kKeyShare1, aes_cfg->key_len);
+  return kSimpleSerialOk;
+}
+
+/*
+ * Simple serial trigger AES encryption command.
+ *
+ * @param aes_cfg        AES configuration object.
+ * @param plain_text     plain text to be encrypted.
+ * @param plain_text_len plain text length.
+ *
+ * @return kSimpleSerialOk on success, kSimpleSerialError otherwise.
+ */
+static simple_serial_result_t simple_serial_trigger_encryption(
+    const char *plain_text, size_t plain_text_len) {
+  if (plain_text_len > kMaxInputLengthBin) {
+    return kSimpleSerialError;
+  }
+
+  char cipher_text[64];
+  size_t num_blocks = get_num_blocks(plain_text_len);
+  if (num_blocks != 1) {
+    return kSimpleSerialError;
+  }
+
+  // start timer to wake up Ibex after AES is done.
+  uint64_t current_time;
+  CHECK(dif_rv_timer_counter_read(&timer, kHart, &current_time) ==
+        kDifRvTimerOk);
+  CHECK(dif_rv_timer_arm(&timer, kHart, kComparator,
+                         current_time + kDeadline) == kDifRvTimerOk);
+  CHECK(dif_rv_timer_counter_set_enabled(&timer, kHart, kDifRvTimerEnabled) ==
+       kDifRvTimerOk);
+
+  aes_data_put_wait(plain_text);
+  if (dif_gpio_all_write(&gpio, kGpioCaptureTriggerH) != kDifGpioOk) {
```

Comment:
Is there a reason why some checks use `CHECK` while others return `kSimpleSerialError`? If so, would it make sense to have another macro like that calls `CHECK` but also returns `kSimpleSerialError` to log these cases as well?

Reply:
Added `SS_CHECK` to make the checks more consistent. Once the capture starts, we should comply with the simple serial protocol, which as a limited buffer at the moment. So, only returning `kSimpleSerialError` until we update we decide if we want to update the host side.